win32/devdoc/file_win32_requirements.md

```diff
@@ -0,0 +1,133 @@
+
+**SRS_FILE_WIN32_43_028: [** `file_read_async` shall populate the `ov` field of the `FILE_WIN32_IO` using `position`. **]**
+
+**SRS_FILE_WIN32_43_029: [** `file_read_async` shall call `ReadFile`. **]**
+
+**SRS_FILE_WIN32_43_030: [** If `ReadFile` fails synchronously and `GetLastError` indicates `ERROR_IO_PENDING` then `file_read_async` shall succeed and return `FILE_READ_ASYNC_OK`. **]**
+
+**SRS_FILE_WIN32_43_031: [** If `ReadFile` fails synchronously and `GetLastError` does not indicate `ERROR_IO_PENDING` then `file_read_async` shall fail, call `CancelThreadpoolIo` and return `FILE_READ_ASYNC_WRITE_ERROR`. **]**
+
+**SRS_FILE_WIN32_43_032: [** If `ReadFile` succeeds synchronously then `file_read_async` shall succeed, call `CancelThreadpoolIo` and `user_callback` and return `FILE_READ_ASYNC_OK`. **]**
+
+
+## on_file_io_complete_win32
+
+```c
+static VOID CALLBACK on_file_io_complete_win32( /*called when some read/write operation is finished*/
+    _Inout_     PTP_CALLBACK_INSTANCE Instance,
+    _Inout_opt_ PVOID                 Context,
+    _Inout_opt_ PVOID                 Overlapped,
+    _In_        ULONG                 IoResult,
+    _In_        ULONG_PTR             NumberOfBytesTransferred,
+    _Inout_     PTP_IO                Io
+);
+```
+`on_file_io_complete_win32` is called when a file operation completes asynchronously.
+
+
+**SRS_FILE_WIN32_43_034: [** `on_file_io_complete_win32` shall recover the context containing `Overlapped`. **]**
+
+**SRS_FILE_WIN32_43_035: [** If `IoResult` is not `NO_ERROR`, `on_file_io_complete_win32` shall call `user_report_fault_callback` with `user_report_fault_context` which were specified when `file_create` was called. **]**
+
+**SRS_FILE_WIN32_43_036: [** If the type of the asynchronous operation is read, `on_file_io_complete_win32` shall  construct a `CONSTBUFFER_HANDLE` by calling `CONSTBUFFER_CreateWithMoveMemory` from the bytes read by `ReadFile`. **]**
+
+**SRS_FILE_WIN32_43_037: [** If the construction of the `CONSTBUFFER_HANDLE` fails, `on_file_io_complete_win32` shall call `user_callback` with `user_context`, `false` as `is_successful` and `NULL` as `content`. **]**
+
+**SRS_FILE_WIN32_43_038: [** If the construction of the `CONSTBUFFER_HANDLE` succeeds, `on_file_io_complete_win32` shall call `user_callback` with `user_context`, `true` as `is_successful` and the created `CONSTBUFFER_HANDLE` as `content`. **]**
+
+**SRS_FILE_WIN32_43_039: [** If the type of the asynchronous operation is write, `on_file_io_complete_win32` shall call `user_callback` with `user_context` and `true` as `is_successful`. **]**
+
+**SRS_FILE_WIN32_43_040: [** If the type of asynchronous operation is neither read nor write, `on_file_io_complete_win32` shall call `user_report_fault_callback` with `user_report_fault_context` which were specified when `file_create` was called. **]**
```

Comment:
>If the type of asynchronous operation is neither read nor write [](start = 29, length = 63)

This isn't really testable, so you should do this, but there's really no spec for it